### PR TITLE
[BYOM] Custom System Prompts

### PR DESCRIPTION
**Pull request by jonathansampson**
Introduces support for custom system prompts for BYOM models. When set, the custom system prompt will replace Brave's default prompt at the start of each conversation.

Also adds support for an optional %datetime% placeholder, which (if encountered) will be replaced with the same date/time string used in Brave's standard system prompt.

Resolves https://github.com/brave/brave-browser/issues/40690
Security Review https://github.com/brave/reviews/issues/1759

image

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Create a custom model within brave://settings/leo-assistant
2. Assign your custom system prompt
3. Verify that your prompt is present in outbound request
  a. can be achieved either by observing network activity, or
  b. by using an easily verifiable prompt (e.g., "You answer only in all caps.") with a reasonably capable model (e.g., Phi3.5)

### CODE REVIEW

File: app/brave_settings_strings.grdp [hunk 1]
```diff
@@ -1411,6 +1411,12 @@ Leo. This action can't be undone.
   <message name="IDS_SETTINGS_LEO_ASSISTANT_SAVE_MODEL_BUTTON_LABEL" desc="A button label for user to click when they want to save a new model">
     Save model
   </message>
+  <message name="IDS_SETTINGS_LEO_ASSISTANT_MODEL_SYSTEM_PROMPT_TITLE" desc="A header title for the model system prompt">
+    System Prompt
+  </message>
+  <message name="IDS_SETTINGS_LEO_ASSISTANT_MODEL_SYSTEM_PROMPT_DESC" desc="A description message for the custom model system prompt">
+    Custom instructions and/or context given to the model to guide its responses in a specific direction or style. Use &#x25;datetime&#x25; where you would like the current date and time to be inserted.
```
Comment (fallaciousreasoning) [inline app/brave_settings_strings.grdp:1418]: should we mention that this is optional and that Brave provides a default prompt if it isn't specified?

Comment (jonathansampson) [reply app/brave_settings_strings.grdp:1418]: Good call, @fallaciousreasoning! I wonder if we should pre-populate the field with our own default prompt. Then the user can choose to keep it, or modify. Might be even more user-friendly than simply presenting them with an empty textarea...

File: browser/resources/settings/brave_leo_assistant_page/model_config_ui.html [hunk 2]
```diff
@@ -174,6 +174,36 @@
...
+              class="tooltip-content"
+            >
+              $i18n{braveLeoAssistantModelSystemPromptDesc}
+            </div>
+            <leo-button
+              size="small"
+              kind="plain-faint"
+              fab
+            >
+              <cr-icon class="small"  icon="info-outline"></cr-icon>
```
Comment (fallaciousreasoning) [inline browser/resources/settings/brave_leo_assistant_page/model_config_ui.html:194]: could you try using a `leo-icon` here instead - I've noticed some of the sizing is a bit funky with a `cr-icon` focus ring:


```suggestion
              <leo-icon name="info-outline"></cr-icon>
```

(size will hopefully be implied by the button size)

File: components/ai_chat/core/browser/engine/engine_consumer_oai.cc [hunk 3]
```diff
@@ -54,9 +55,18 @@ base::Value::List BuildMessages(
   {
     std::string date_and_time_string =
         base::UTF16ToUTF8(TimeFormatFriendlyDateAndTime(base::Time::Now()));
-    std::string system_message = base::StrCat({base::ReplaceStringPlaceholders(
-        l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_SYSTEM_MESSAGE_GENERIC),
-        {date_and_time_string}, nullptr)});
+    std::string system_message;
+
+    if (!model_options.model_system_prompt.empty()) {
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/engine/engine_consumer_oai.cc:60]: I wonder if a `std::optional` makes more sense here?

File: components/ai_chat/core/browser/model_service.cc [hunk 4]
```diff
@@ -259,6 +263,26 @@ void ModelService::MigrateProfilePrefs(PrefService* profile_prefs) {
...
+    bool models_updated = false;
+
+    for (auto& model_value : custom_models_list) {
+      if (base::Value::Dict* model_dict = model_value.GetIfDict()) {
+        if (!model_dict->FindString(kCustomModelSystemPromptKey)) {
+          model_dict->Set(kCustomModelSystemPromptKey, "");
+          models_updated = true;
+        }
+      }
+    }
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/model_service.cc:281]: optional: maybe it makes sense to allow null (or `std::nullopt`) here, as conceptually the empty string and `null` feel different (one is that there is no custom prompt, and the other is that the custom prompt is empty)

Also, we wouldn't need this migration code

**Review by fallaciousreasoning — APPROVED** [review]
lgtm

File: app/brave_settings_strings.grdp (hunk 1 above)
Comment (jonathansampson) [reply app/brave_settings_strings.grdp:1418]: @fallaciousreasoning Thoughts on this approach? I've moved away from using the LLAMA2 prompt, and instead created a dedicated prompt-string for custom models. The first four lines of the prompt are visible in the prompt textarea, demonstrating a default prompt to the user, as well as use of the %datetime% placeholder.

image

File: browser/resources/settings/brave_leo_assistant_page/model_config_ui.html (hunk 2 above)
Comment (jonathansampson) [reply browser/resources/settings/brave_leo_assistant_page/model_config_ui.html:194]: Looks like the leo-icon version is positioned a bit higher than the cr-icon elements. 

leo-icon:

image

cr-icon:

image

File: components/ai_chat/core/browser/engine/engine_consumer_oai.cc (hunk 3 above)
Comment (jonathansampson) [reply components/ai_chat/core/browser/engine/engine_consumer_oai.cc:60]: @fallaciousreasoning Can you elaborate? The `system_message` _will be_ populated, either in the first condition block (user-defined prompt) or by the else-block (default prompt).

File: browser/resources/settings/brave_leo_assistant_page/model_config_ui.html (hunk 2 above)
Comment (fallaciousreasoning) [reply browser/resources/settings/brave_leo_assistant_page/model_config_ui.html:194]: Ah feel free for either then :)

File: components/ai_chat/core/browser/engine/engine_consumer_oai_unittest.cc [hunk 5]
```diff
@@ -196,12 +198,98 @@ TEST_F(EngineConsumerOAIUnitTest, GenerateQuestionSuggestions) {
...
+      ));
+
+  // Prepare to capture API client request
+  auto* client = GetClient();
+  auto run_loop = std::make_unique<base::RunLoop>();
+
+  // Expect a single call to PerformRequest
+  EXPECT_CALL(*client, PerformRequest(_, _, _, _))
+      .WillOnce(
+          [&](const mojom::CustomModelOptions, base::Value::List messages,
```
Comment (darkdh) [inline components/ai_chat/core/browser/engine/engine_consumer_oai_unittest.cc:254]: nit: only capture the things you need

**Review by darkdh — APPROVED** [review]

**Comment by jonathansampson** [comment]
There's a bug right now which may be in cr-textarea's implementation. When you click the ✏️ icon to edit a model, you are sometimes required to scroll down to find the edit form. Other times, the form is rendered beneath your mouse cursor, such that you are hovering the System Prompt input. Importantly, your mouse never _moved into the System Prompt textarea_, it was simply given position atop the textarea thanks to a bit of script aiming to make the form more easily accessed. When this happens, if you click on the textarea, the page will crash. If you leave the textarea, and then re-enter it before clicking, no crash will occur.

https://github.com/user-attachments/assets/79f4e5d9-5ad6-4acb-b9c0-ccc9cf5a99e6

**Review by mkarolin — APPROVED** [review]
`strings`++

**Comment by awaescher** [comment]
Just to give a little feedback: I was looking forward to this since weeks and just noticed the update.
Thanks a thousand times, Brave is such a great browser. Really appreciate your work :heart:

Image

**Pull request merged**